### PR TITLE
[SPARK-35839][SQL] New SQL function: to_timestamp_ntz

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Implement new SQL function: `to_timestamp_ntz`. 
The syntax is similar to the built-in function `to_timestamp`:
```
to_timestamp_ntz ( <date_expr> )

to_timestamp_ntz ( <timestamp_expr> )

to_timestamp_ntz ( <string_expr> [ , <format> ] )
```

The naming is from snowflake: https://docs.snowflake.com/en/sql-reference/functions/to_timestamp.html

### Why are the changes needed?

Adds a new SQL function to create a literal/column of timestamp without time zone.
It's convenient for both end-users and developers.

### Does this PR introduce _any_ user-facing change?

Yes, a new SQL function `to_timestamp_ntz`.

### How was this patch tested?

Unit tests

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/ansi/datetime.sql.out [hunk 1]
```diff
@@ -936,6 +1198,33 @@ org.apache.spark.SparkUpgradeException
 You may get a different result due to the upgrading of Spark 3.0: Fail to recognize 'dd MM yyyy EEEEE' pattern in the DateTimeFormatter. 1) You can set spark.sql.legacy.timeParserPolicy to LEGACY to restore the behavior before Spark 3.0. 2) You can form a valid datetime pattern with the guide from https://spark.apache.org/docs/latest/sql-ref-datetime-pattern.html
 
 
+-- !query
+select to_timestamp_ntz('2019-10-06 A', 'yyyy-MM-dd GGGGG')
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.SparkUpgradeException
+You may get a different result due to the upgrading of Spark 3.0: Fail to recognize 'yyyy-MM-dd GGGGG' pattern in the DateTimeFormatter. 1) You can set spark.sql.legacy.timeParserPolicy to LEGACY to restore the behavior before Spark 3.0. 2) You can form a valid datetime pattern with the guide from https://spark.apache.org/docs/latest/sql-ref-datetime-pattern.html
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/results/ansi/datetime.sql.out:1207]: I will have a follow up to improve the error message.
This Error message is used in multiple SQL output files and some of them are not related to this PR.

**Comment by MaxGekk** [comment]
@gengliangwang Thanks for the ping. I will review it today slightly later.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 2]
```diff
@@ -379,6 +379,7 @@ object SparkExecuteStatementOperation {
         case CalendarIntervalType => StringType.catalogString
         case _: YearMonthIntervalType => "interval_year_month"
         case _: DayTimeIntervalType => "interval_day_time"
+        case _: TimestampWithoutTZType => "timestamp"
```
Comment (gengliangwang) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:382]: It seems that there is not better choice from: https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/thrift/Type.java#L230

**Comment by cloud-fan** [comment]
2 high-level comments:
1. I think this new function should just go with the non-legacy behavior if the legacy conf is enabled, instead of failing.
2. We can make `ToTimestamp` more general to support this new function, instead of adding a new subclass of `ToTimestamp`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala [hunk 3]
```diff
@@ -50,9 +51,31 @@ sealed trait TimestampFormatter extends Serializable {
...
+   * @throws DateTimeParseException can be thrown by new parser
+   * @throws DateTimeException unable to obtain local date or time
+   * @throws UnsupportedOperationException cannot use legacy formatter for the parsing
+   */
+  @throws(classOf[ParseException])
+  @throws(classOf[DateTimeParseException])
+  @throws(classOf[DateTimeException])
+  @throws(classOf[UnsupportedOperationException])
+  def parseWithoutTimeZone(s: String): Long =
+    throw QueryExecutionErrors.cannotUseLegacyFormatterForTimestampWithoutTZ();
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala:69]: The error here won't be hit since the formatter for TimestampWithoutTZ is always Iso8601TimestampFormatter

**Comment by gengliangwang** [comment]
@cloud-fan your comments are addressed. PTAL.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 4]
```diff
@@ -66,6 +66,10 @@ trait TimestampFormatterHelper extends TimeZoneAwareExpression {
 
   protected def isParsing: Boolean
 
+  // Whether the timestamp formatter is for TimestampWithoutTZType.
+  // If yes, the formatter is always `Iso8601TimestampFormatter`.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:70]: Shall we put this logic in `TimestampFormatterHelper`?
```
final protected def getFormatter(fmt: String): TimestampFormatter = {
  if (forTimestampWithoutTZ) new Iso8601TimestampFormatter(...) else ...
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 5]
```diff
@@ -995,6 +1000,77 @@ case class UnixTimestamp(
     copy(timeExp = newLeft, format = newRight)
 }
 
+case class GetTimestampWithoutTZ(
+    left: Expression,
+    right: Expression,
+    timeZoneId: Option[String] = None,
+    failOnError: Boolean = SQLConf.get.ansiEnabled) extends ToTimestamp {
+
+  override lazy val forTimestampWithoutTZ: Boolean = true
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:1009]: it's a constant, no need to do lazy val.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 6]
```diff
@@ -1013,7 +1089,7 @@ abstract class ToTimestamp
   override def dataType: DataType = LongType
   override def nullable: Boolean = if (failOnError) children.exists(_.nullable) else true
 
-  private def isParseError(e: Throwable): Boolean = e match {
+  protected def isParseError(e: Throwable): Boolean = e match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:1092]: unnecessary change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala:69]: This is not a user-facing error: we should never hit it. We can just throw IllegelStateException here and no need to document this with `@throws`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala [hunk 7]
```diff
@@ -50,9 +51,31 @@ sealed trait TimestampFormatter extends Serializable {
...
+
   def format(us: Long): String
   def format(ts: Timestamp): String
   def format(instant: Instant): String
+
+  @throws(classOf[UnsupportedOperationException])
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala:75]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:70]: Well I prefer to put all the logic in `TimestampFormatter.getFormatter`

File: sql/core/src/test/resources/sql-tests/results/ansi/datetime.sql.out (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/ansi/datetime.sql.out:1223]: good point. The new function should bypass this upgrade check in the timestamp formatter.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:382]: We should probably map the existing timestamp to `TIMESTAMPLOCALTZ`

Comment (gengliangwang) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:382]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 8]
```diff
@@ -1225,6 +1225,12 @@ object QueryExecutionErrors {
     new RuntimeException("Remote operations not supported")
   }
 
+  def missingMethodForTimestampWithoutTZ(method: String): Throwable = {
+    new IllegalStateException(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1229]: This is not a user-facing error and we should just inline it, similar to `UnresolvedAttribute.exprId`

**Review by cloud-fan — APPROVED** [review]
LGTM except for one comment

**Comment by gengliangwang** [comment]
Merging to master
